lib/block_view_atomic_txns.go

```diff
@@ -0,0 +1,532 @@
+		}
+
+		// Convert the txnBytes buffer to a MsgDeSoTxn struct.
+		if err = txnData.Txns[ii].FromBytes(txnBytes); err != nil {
+			return errors.Wrap(err,
+				"AtomicTxnsWrapperMetadata.FromBytes: Problem parsing transaction bytes")
+		}
+	}
+	return nil
+}
+
+func (txnData *AtomicTxnsWrapperMetadata) New() DeSoTxnMetadata {
+	return &AtomicTxnsWrapperMetadata{}
+}
+
+//
+// HELPER FUNCTIONS: MsgDeSoTxn
+//
+
+// IsAtomicTxnsInnerTxn is used to determine if a MsgDeSoTxn is an inner txn
+// of an atomic transaction. An atomic transaction is qualified by the existence
+// of the NextAtomicTxnPreHash and PreviousAtomicTxnPreHash keys in the ExtraData map.
+func (msg *MsgDeSoTxn) IsAtomicTxnsInnerTxn() bool {
+	if _, keyExists := msg.ExtraData[NextAtomicTxnPreHash]; !keyExists {
+		return false
+	}
+	if _, keyExists := msg.ExtraData[PreviousAtomicTxnPreHash]; !keyExists {
+		return false
+	}
+	return true
+}
+
+// AtomicHash calculates the "atomic" hash of a MsgDeSoTxn by removing the
+// NextAtomicTxnPreHash and PreviousAtomicTxnPreHash keys in the ExtraData
+// map as well as the transaction signature and computing the SHA256 double
+// hash of the resulting transaction.
+//
+// The atomic hash is only meant for use in validating the sequence of
+// a series of atomic transactions. Transactions are sequenced by
+// referencing each-other's atomic hashes in their ExtraData map under the
+// NextAtomicTxnPreHash and PreviousAtomicTxnPreHash. For a complete explanation
+// on how to create a series of atomic transactions, read the
+// AtomicTxnsWrapperMetadata comment.
+func (msg *MsgDeSoTxn) AtomicHash() (*BlockHash, error) {
+	// Create a duplicate of the transaction to ensure we don't edit the existing transaction.
+	msgDuplicate, err := msg.Copy()
+	if err != nil {
+		return nil, errors.Wrap(err, "MsgDeSoTxn.AtomicHash: Cannot create duplicate transaction")
+	}
+
+	// Delete the NextAtomicTxnPreHash and PreviousAtomicTxnPreHash from the ExtraData map.
+	delete(msgDuplicate.ExtraData, NextAtomicTxnPreHash)
+	delete(msgDuplicate.ExtraData, PreviousAtomicTxnPreHash)
+
+	// Convert the transaction to bytes but do NOT encode the transaction signature.
+	txBytes, err := msgDuplicate.ToBytes(true)
+	if err != nil {
+		return nil, errors.Wrap(err, "MsgDeSoTxn.AtomicHash: cannot convert modified transaction to bytes")
+	}
+
+	// Return the SHA256 double hash of the resulting bytes.
+	return Sha256DoubleHash(txBytes), nil
+}
+
+//
+// Connect and Disconnect Atomic Txn Logic
+//
+
+func (bav *UtxoView) _connectAtomicTxnsWrapper(
+	txn *MsgDeSoTxn,
+	txHash *BlockHash,
+	blockHeight uint32,
+	blockTimestampNanoSecs int64,
+	verifySignatures bool,
+	ignoreUtxos bool,
+) (
+	_utxoOps []*UtxoOperation,
+	_totalInput uint64,
+	_totalOutput uint64,
+	_fees uint64,
+	_err error,
+) {
+	var utxoOpsForTxn []*UtxoOperation
+
+	// Validate the connecting block height.
+	if blockHeight < bav.Params.ForkHeights.ProofOfStake1StateSetupBlockHeight {
```

Comment:
If atomic txns are gated by `ProofOfStake1StateSetupBlockHeight`, then they need to be supported by the old PoW mempool. @poolcoke, @lazynina - has that been regtested yet? NBD if not. We'll just need to remember to regtest it.